bpf/lib/ipip_termination.h

```diff
@@ -0,0 +1,78 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+
+
+#ifndef __LIB_IPIP_TERMINATION_H_
+#define __LIB_IPIP_TERMINATION_H_
+
+#ifdef ENABLE_IPIP_TERMINATION
+
+static __always_inline int
+decap_ipv4(struct __ctx_buff *ctx)
+{
+	struct iphdr *outer_ip4;
+	struct iphdr inner_ip4;
+	void *data, *data_end;
+
+	struct ipv4_ct_tuple tuple = {};
+	int ret, inner_ip4_off, outer_ip4_off = ETH_HLEN, outer_ip4_len, l4_off;
+
+	struct lb4_key key = {};
+	struct lb4_service *svc;
+
+	if (!revalidate_data_pull(ctx, &data, &data_end, &outer_ip4))
+		return DROP_INVALID;
+	if (outer_ip4->protocol != IPPROTO_IPIP) {
+		return CTX_ACT_OK;
+	}
+
+	outer_ip4_len = ipv4_hdrlen(outer_ip4);
+	inner_ip4_off = outer_ip4_off + outer_ip4_len;
+
+	if (ctx_load_bytes(ctx, inner_ip4_off, &inner_ip4, sizeof(inner_ip4)) < 0)
+		return DROP_INVALID;
+
+	ret = lb4_extract_tuple(ctx, &inner_ip4, inner_ip4_off, &l4_off, &tuple);
+	if (IS_ERR(ret)) {
+		/* Bypass */
+		return CTX_ACT_OK;
+	}
+
+	lb4_fill_key(&key, &tuple);
+
+	svc = lb4_lookup_service(&key, false);
+	if (svc) {
+		if (!lb4_svc_is_loadbalancer(svc))
+			return CTX_ACT_OK;
+
+		if (ctx_adjust_hroom(ctx, -outer_ip4_len, BPF_ADJ_ROOM_MAC, BPF_F_ADJ_ROOM_FIXED_GSO))
```

Comment:
There is no specific reason to use `BPF_F_ADJ_ROOM_FIXED_GSO`. Using `ctx_adjust_hroom_flags()` works correctly. And I've verified its proper operation over the past year in our cloud. However, in `test_tc_tunnel.c` `BPF_F_ADJ_ROOM_FIXED_GSO` is used when performing IPIP decapsulation. I included this flag in case there were any old issues that I might not be aware of.
https://github.com/cilium/cilium/blob/10b3a73a0198139b0e51a7179953c5647268e062/test/l4lb/test_tc_tunnel.c#L495


From the kernel code perspective, there is no difference between using `BPF_F_ADJ_ROOM_FIXED_GSO` and `ctx_adjust_hroom_flags()` in this context.

https://github.com/torvalds/linux/blob/4376e966ecb78c520b0faf239d118ecfab42a119/net/core/filter.c#L3704-L3707
```C
	ret = shrink ? bpf_skb_net_shrink(skb, off, len_diff_abs, flags) :
		       bpf_skb_net_grow(skb, off, len_diff_abs, flags);
	if (!ret && !(flags & BPF_F_ADJ_ROOM_NO_CSUM_RESET))
		__skb_reset_checksum_unnecessary(skb);
```

It would be better to change this code to use `ctx_adjust_hroom_flags()` as well.

Reply:
https://github.com/cilium/cilium/blob/27b50319b1a2767aeac8efef6d04d81588c46008/bpf/lib/ipip_termination.h#L44-L45